bayes_opt1/util.py

```diff
@@ -0,0 +1,547 @@
+        '''
+        gp.fit(x_obs, y_obs)
+        mu = gp.predict(grid, return_std=False)
+        return mu
+    
+    def find_mean_max(x_obs,y_obs,gp,grid,mean=None,std=None,i=None):
+        '''
+        If i is None, then it means we are looking for what the paper called \mu^{*}_{n}, otherwise we want to
+        compute \mu^{*}_{n+1}.
+        Remark: i=None <=> mean=None <=> std=None
+        '''
+        if isinstance(i, type(None)): # checks if i is None
+            x_newobs = x_obs
+            y_newobs = y_obs
+        else:
+            y=np.random.normal(loc=mean, scale=std)
+            x_newobs=np.concatenate((x_obs, i.reshape(1,-1)), axis = 0)
+            y_newobs=np.append(y_obs,y) 
+        mu=posterior(gp, x_newobs, y_newobs, grid)
+        return (max(mu),np.argmax(mu))
+    
+    p = optimizer._space.bounds.shape[0] # size of state space (e.g. if we are working on R^3, n = 3)
+    init = [np.zeros(n_grid)] * p # list of p vectors, each of lenght n_grid   
+    x_obs_temp = np.array([[res["params"]] for res in optimizer.res]) 
+    y_obs = np.array([res["target"] for res in optimizer.res])
+    # x_obs needs to be np.array of size (n,p)
+    # y_obs needs to be np.array of size (n,)
+    # see https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.GaussianProcessRegressor.html
+    x_obs = x_obs_to_array(x_obs_temp, p)
+    
+
+    for i in range(p): 
+        init[i] = np.linspace(optimizer._space.bounds[i,0], optimizer._space.bounds[i,1], n_grid) # works \forall p
+
+    grid = np.meshgrid(*init) # list with p matrices n_grid x n_grid
+    for g in range(len(grid)):
+        grid[g] = grid[g].reshape(-1,1)
+    grid = np.stack(grid, axis = -1)
+    grid=grid[:,0,:] # array of shape (n_grid*n_grid, p): each row is composed of one element from init[0] and one element from init[1]
+                
+    temp = find_mean_max(x_obs,y_obs,gp,grid)
+    mean_nstar = temp[0]
+
+    ii = 0
+    
+    if p == 1:
+        x = np.array([x])
+        
+    diff = np.zeros(len(x))
+    for i in x:
+        print(ii+1)
+        mean, std= gp.predict(i.reshape(1,-1), return_std=True)
+        for j in range(J):
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore")
+                temp = find_mean_max(x_obs,y_obs,gp,grid,mean,std,i)
+                mean_n1star = temp[0]
+                diff[ii] = diff[ii] + (mean_n1star-mean_nstar)
+    #             print( ((ii + 1)/len(x)*100 , (j+1)/J*100) )
+        ii = ii + 1
+    diff = diff/J
+    return diff
+
+
+# Implementation algorithm 4 (Frazier)
+
+def _kg4(x, optimizer, gp, n_grid = 100, J = 300):
+    '''
+    Library "numdifftools" needed:
+    
+                                    pip install numdifftools
+    REMARK: if p == 1, then x needs to be a np.float
+            if p > 1, then x needs to be a np.ndarray with of the form [,]
+    '''
+    p = optimizer._space.bounds.shape[0]
+    
+    if p == 1:
```

Comment:
Come mai avete distinto i casi uni- e multi-dimensionali? Immagino che nel primo caso, costruendo un `np.array` uni-dimensionale si ricada nell'altro... Cosa è andato storto?

Reply:
Per generalità il for successivo è multidimensionale e per usare metodi come .reshape (o anche per iterare nel ciclo) serve avere un array e non uno scalare, quindi per comodità abbiamo reso uno scalare un array.

Reply:
La mia domanda era: perché c'è stato bisogno di un `if dim == 1: cose() else: altrecose()` invece che trattarle con un unico blocco di codice?

Reply:
Perchè volevamo imporre un controllo in modo che l'input avesse un solo punto e abbiamo diviso i due casi dim=1 e non perchè in dimensione 1 l'input potrebbe essere un float, e non un array... abbiamo cambiato il codice togliendo l'else, ma aggiungendo un if che controlla il tipo di input e lo trasforma in un array.